src/task.c

```diff
@@ -848,6 +884,43 @@ int task_lock(struct task *t) {
         if (ci->hydro.hold) return 0;
         if (cell_locktree(ci) != 0) return 0;
 #endif
+      } else if (subtype == task_subtype_gpu_pack) {
+        /* Attempt to lock the cell */
+        if (ci->hydro.hold)
+          return 0;
+        if (cell_locktree(ci) != 0)
+          return 0;
+      } else if (subtype == task_subtype_gpu_pack_f) {
+        /* Attempt to lock the cell */
+        if (ci->hydro.hold)
+          return 0;
+        if (cell_locktree(ci) != 0)
+          return 0;
+      } else if (subtype == task_subtype_gpu_pack_g) {
+        /* Attempt to lock the cell */
+        if (ci->hydro.hold)
+          return 0;
+        if (cell_locktree(ci) != 0)
+          return 0;
+      } else if (subtype == task_subtype_gpu_unpack) {
+//        for(int pp = 0; pp < 128 /*should be sched->pack_size*/; pp++){
+//    	  if (t->ci_unpack[pp]->gpu_done == 0){
+//    		  message("trying to queue an unpack before all packs done on GPU");
+//    		  return 0;
+//    	  }
+////          if (t->ci_unpack[pp]->hydro.hold)
+////    		return 0;
+////    	  if (cell_locktree(t->ci_unpack[pp]) != 0)
+////            return 0;
+//        }
+        /* Nothing to do here */
+        return 1;
+      } else if (subtype == task_subtype_gpu_unpack_f) {
```

Comment:
Why no locking for unpacking? Shouldn't we prevent other threads from writing?

Reply:
The unpacking task does nothing. It's just a mechanism used to ensure that a collection of N tasks which the GPU is working on have all been completed before we move onto their dependencies. It is only unlocked when all N tasks have been completed and all the unpack task does is unlock the correct dependencies for the GPU tasks